File: DAMOS/trunk/tests/org.eclipselabs.damos.codegen.c.tests/src/org/eclipselabs/damos/codegen/c/gtests/ArrayElementWiseOperationTest.java
/****************************************************************************
 * Copyright (c) 2008, 2012 Andreas Unger and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *    Andreas Unger - initial API and implementation 
 ****************************************************************************/

package org.eclipselabs.damos.codegen.c.gtests;

import org.eclipselabs.damos.codegen.c.test.AbstractGeneratorGTest;
import org.eclipselabs.damos.codegen.c.test.GTest;
import org.eclipselabs.damos.dml.Block;
import org.eclipselabs.damos.dml.Inport;
import org.eclipselabs.damos.dml.Outport;
import org.junit.Before;

/**
 * @author Andreas Unger
 *
 */
@GTest(sourceFile="gtests/ArrayElementWiseOperationTest.cpp", program="gtests/ArrayElementWiseOperationTest")
public class ArrayElementWiseOperationTest extends AbstractGeneratorGTest {

	@Before
	@Override
	public void setUp() {
		super.setUp();
		
		Inport inA = createInport("InA", createIntegerTypeSpecification(), 1);
		Inport inB = createInport("InB", createIntegerTypeSpecification(), 1);
		Block arrayTest = createTestBlock("ArrayElementWiseOperationTest", "ArrayElementWiseOperationTest");
		Outport outport1 = createOutport("Out1", createDataTypeSpecification("real[3]"));
		Outport outport2 = createOutport("Out2", createDataTypeSpecification("real[3, 2]"));
		
		connect(inA, arrayTest, 0);
		connect(inB, arrayTest, 1);
		connect(arrayTest, 0, outport1);
		connect(arrayTest, 1, outport2);

		generateAndCompile();
	}

}

File: DAMOS/branches/initial-contribution/tests/org.eclipse.damos.codegen.c.tests/gtests/ArrayElementWiseOperationTest.cpp
/*******************************************************************************
 * Copyright (c) 2008, 2012 Andreas Unger and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *     Andreas Unger - initial API and implementation
 *******************************************************************************/
#include "gtest/gtest.h"
#include "ArrayElementWiseOperationTest.h"

TEST(ArrayElementWiseOperationTest, ArrayElementWiseOperation) {
	Input input;
	Output output;

	initialize();

	input.inA = 4;
	input.inB = 3;

	execute(&input, &output);

	EXPECT_DOUBLE_EQ(7.0, output.out1[0]);
	EXPECT_DOUBLE_EQ(-0.9, output.out1[1]);
	EXPECT_DOUBLE_EQ(24.0, output.out1[2]);

	EXPECT_DOUBLE_EQ(7, output.out2[0][0]);
	EXPECT_DOUBLE_EQ(-0.9, output.out2[0][1]);
	EXPECT_DOUBLE_EQ(30, output.out2[1][0]);
	EXPECT_DOUBLE_EQ(40, output.out2[1][1]);
	EXPECT_DOUBLE_EQ(51, output.out2[2][0]);
	EXPECT_DOUBLE_EQ(62, output.out2[2][1]);
}
